Build error during build from source on Sonoma M1 Max

### Checklist

- [X] I have searched for [similar issues](https://github.com/isl-org/Open3D/issues).
- [X] For Python issues, I have tested with the [latest development wheel](http://www.open3d.org/docs/latest/getting_started.html#development-version-pip).
- [X] I have checked the [release documentation](http://www.open3d.org/docs/release/) and the [latest documentation](http://www.open3d.org/docs/latest/) (for `master` branch).

### Steps to reproduce the issue

#### I first cloned Open3D by:
```
git clone https://github.com/isl-org/Open3D.git
cd Open3D
```

#### Then, I build Open3D (on Ubuntu 20.04, with CUDA 11.5) with:
```
mkdir build && cd build
cmake .. -DBUILD_PYTHON_MODULE=OFF 
make -j$(sysctl -n hw.physicalcpu)
```


### Error message

```shell
[ 84%] Built target core
/Open3Ds/open3dlatest/cpp/open3d/pipelines/registration/Feature.cpp:158:15: error: unused variable 'kOuterThreads' [-Werror,-Wunused-variable]
    const int kOuterThreads = std::min(kMaxThreads, num_searches);
              ^
/Open3Ds/open3dlatest/cpp/open3d/pipelines/registration/Feature.cpp:159:15: error: unused variable 'kInnerThreads' [-Werror,-Wunused-variable]
    const int kInnerThreads = std::max(kMaxThreads / num_searches, 1);
              ^
2 errors generated.
make[2]: *** [cpp/open3d/pipelines/CMakeFiles/pipelines.dir/registration/Feature.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
/Open3Ds/open3dlatest/cpp/open3d/t/pipelines/registration/Feature.cpp:104:15: error: unused variable 'kOuterThreads' [-Werror,-Wunused-variable]
    const int kOuterThreads = std::min(kMaxThreads, num_searches);
              ^
1 error generated.
```


### Open3D, Python and System information

```markdown
- Operating system: macOS 14.1.1
- Python version: Doesn't matter ( I think :D )
- Open3D version: tried 17.0 and latest from master branch
- System architecture: apple-silicon (M1 Max)
- Is this a remote workstation?: No
- How did you install Open3D?: build from source
- Compiler version (if built from source): Clang 14 (from Xcode)
```


### Additional information

So yeah, firstly I got error during building VTK, but then got this. Now will surpress those warnings and then provide more comments

--- a/cpp/open3d/pipelines/registration/Feature.cpp
+++ b/cpp/open3d/pipelines/registration/Feature.cpp
@@ -157,6 +157,8 @@ CorrespondenceSet CorrespondencesFromFeatures(const Feature &source_features,
 
     const int kOuterThreads = std::min(kMaxThreads, num_searches);
     const int kInnerThreads = std::max(kMaxThreads / num_searches, 1);
+    (void)kOuterThreads;
+    (void)kInnerThreads;
 #pragma omp parallel for num_threads(kOuterThreads)
     for (int k = 0; k < num_searches; ++k) {
         geometry::KDTreeFlann kdtree(features[1 - k]);

--- a/cpp/open3d/pipelines/registration/Feature.cpp
+++ b/cpp/open3d/pipelines/registration/Feature.cpp
@@ -154,10 +154,9 @@ CorrespondenceSet CorrespondencesFromFeatures(const Feature &source_features,
     std::vector<CorrespondenceSet> corres(num_searches);
 
     const int kMaxThreads = utility::EstimateMaxThreads();
-
     const int kOuterThreads = std::min(kMaxThreads, num_searches);
     const int kInnerThreads = std::max(kMaxThreads / num_searches, 1);
-    (void)kOuterThreads;
+    (void)kOuterThreads;  // Avoids compiler warning if OpenMP is disabled
     (void)kInnerThreads;
 #pragma omp parallel for num_threads(kOuterThreads)
     for (int k = 0; k < num_searches; ++k) {

--- a/cpp/open3d/t/pipelines/registration/Feature.cpp
+++ b/cpp/open3d/t/pipelines/registration/Feature.cpp
@@ -102,6 +102,7 @@ core::Tensor CorrespondencesFromFeatures(const core::Tensor &source_features,
 
     const int kMaxThreads = utility::EstimateMaxThreads();
     const int kOuterThreads = std::min(kMaxThreads, num_searches);
+    (void)kOuterThreads;
 
     // corres[0]: corres_ij, corres[1]: corres_ji
 #pragma omp parallel for num_threads(kOuterThreads)

--- a/cpp/open3d/t/pipelines/registration/Feature.cpp
+++ b/cpp/open3d/t/pipelines/registration/Feature.cpp
@@ -102,7 +102,7 @@ core::Tensor CorrespondencesFromFeatures(const core::Tensor &source_features,
 
     const int kMaxThreads = utility::EstimateMaxThreads();
     const int kOuterThreads = std::min(kMaxThreads, num_searches);
-    (void)kOuterThreads;
+    (void)kOuterThreads;  // Avoids compiler warning if OpenMP is disabled
 
     // corres[0]: corres_ij, corres[1]: corres_ji
 #pragma omp parallel for num_threads(kOuterThreads)

